core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -365,7 +367,16 @@ class GroupCoordinator(val brokerId: Int,
 
       case Some(group) =>
         group.inLock {
-          if (group.is(Dead) || !group.has(memberId)) {
+          // if a pending member is leaving, it needs to be removed from the pending list, heartbeat cancelled
+          // and if necessary, prompt a JoinGroup completion.
+          if (group.isPendingMember(memberId)) {
+            group.removePendingMember(memberId)
+            heartbeatPurgatory.cancelForKey(MemberKey(groupId, memberId))
```

Comment:
Could we reuse `removeHeartbeatForLeavingMember`? We might need to add the `isLeaving` check to `tryCompleteHeartbeat`.